File: ai-butler-service/scripts/check_markdown_links.py
from __future__ import annotations

import re
import sys
from pathlib import Path
from urllib.parse import unquote

ROOT = Path(__file__).resolve().parents[1]
LINK = re.compile(r"\[[^]]*]\(([^)]+)\)")


def main() -> None:
    failures: list[str] = []
    for document in ROOT.rglob("*.md"):
        if ".venv" in document.parts:
            continue
        for target in LINK.findall(document.read_text(encoding="utf-8")):
            if target.startswith(("http://", "https://", "mailto:", "#")):
                continue
            relative = unquote(target.split("#", maxsplit=1)[0])
            if relative and not (document.parent / relative).resolve().exists():
                failures.append(f"{document.relative_to(ROOT)} -> {target}")
    if failures:
        print("Broken Markdown links:", *failures, sep="\n- ")
        sys.exit(1)
    print("Markdown links are valid")


if __name__ == "__main__":
    main()

File: ai-butler-app/scripts/check-markdown-links.mjs
import { existsSync, readFileSync, readdirSync } from 'node:fs'
import { dirname, extname, join, relative, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'

const root = resolve(dirname(fileURLToPath(import.meta.url)), '..')
const documents = []

function visit(directory) {
  for (const entry of readdirSync(directory, { withFileTypes: true })) {
    if (['.git', 'node_modules', 'dist', 'coverage'].includes(entry.name)) continue
    const path = join(directory, entry.name)
    if (entry.isDirectory()) visit(path)
    else if (extname(entry.name) === '.md') documents.push(path)
  }
}

visit(root)
const failures = []
const link = /\[[^\]]*\]\(([^)]+)\)/g
for (const document of documents) {
  for (const match of readFileSync(document, 'utf8').matchAll(link)) {
    const target = match[1]
    if (!target || /^(https?:|mailto:|#)/.test(target)) continue
    const local = decodeURIComponent(target.split('#')[0])
    if (local && !existsSync(resolve(dirname(document), local))) {
      failures.push(`${relative(root, document)} -> ${target}`)
    }
  }
}
if (failures.length) {
  console.error(`Broken Markdown links:\n- ${failures.join('\n- ')}`)
  process.exit(1)
}
console.log('Markdown links are valid')
